### PR TITLE
[CATROID-1296] Fix NullPointerException in StageLifeCycleController.stageDestroy()

### PR DESCRIPTION
**Pull request by harshsomankar123-tech**
**Description:** This PR addresses a `NullPointerException` (NPE) that occurs during the teardown of `StageActivity` if the `StageListener` hasn't been fully initialized. This was primarily triggered when instrumentation tests failed during the setup/initialization phase, causing the Android runtime to destroy the activity and hit the un-guarded `.finish()` call, which subsequently crashed the entire test process.

**Key Changes:**
* Added a `null` check for `StageActivity.stageListener` in `StageLifeCycleController.stageDestroy()`.
* Added a `null` check for `stageListener` in `StageActivity.manageLoadAndFinish()`.

Reference: https://catrobat.atlassian.net/browse/CATROID-1296

### Checklist for this pull request

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior *(Verified via existing instrumentation tests that previously triggered this crash)*
- [x] Confirm that new and existing unit tests pass locally *(Verified with `connectedCatroidDebugAndroidTest` on emulator)*
- [x] Check that the commits' message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Branch-and-Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow *(Branch: `fix/CATROID-1296-npe-stagedestroy`)*
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java [hunk 1]
```diff
@@ -259,9 +259,11 @@ static void stageDestroy(StageActivity stageActivity) {
 			if (ProjectManager.getInstance().getCurrentProject().isCastProject()) {
 				CastManager.getInstance().onStageDestroyed();
 			}
-			StageActivity.stageListener.finish();
-			stageActivity.manageLoadAndFinish();
-			StageActivity.stageListener = null;
+			if (StageActivity.stageListener != null) {
```
Comment (wslany) [inline catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java]: I think this null-check is a bit too wide. In the crash scenario, when `StageActivity.stageListener` is already `null`, we now also skip `stageActivity.manageLoadAndFinish()`, so the rest of the teardown no longer runs. That means cleanup like `shutDownTextToSpeech()`, speech recognition destruction, Bluetooth pause, and `RaspberryPiService.disconnect()` is bypassed during activity destroy. Would it make sense to guard only the direct `stageListener` calls here and still run the remaining teardown unconditionally? A small regression test for the “listener already null during teardown” path would also help lock this in.

**Review by wslany — CHANGES_REQUESTED** [review]

File: catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java (hunk 1 above)
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java]: @wslany Thanks!
You're right  the null check is too broad and skips important cleanup. I’ll update it to guard only the `stageListener` calls and keep the rest of the teardown running. I’ll also add a regression test for the null case.
and update shortly.

Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java]: Hi @wslany, could you please take a look when you have some time? I’ve completed all the requested changes. Thanks!

Comment (wslany) [reply catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java]: Thanks, this looks good to me now. Guarding only the direct `stageListener.finish()` call fixes the NPE without skipping the rest of teardown, and the added regression test covers the null-listener destroy path nicely. I don’t see any further blockers from my side.`

**Review by wslany — APPROVED** [review]
Looks good to me.

File: catroid/src/test/java/org/catrobat/catroid/stage/StageLifeCycleControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	@Test
+	public void testStageDestroySkipsListenerFinishWhenNull() {
+		StageActivity mockStageActivity = mock(StageActivity.class);
+		ProjectManager mockProjectManager = mock(ProjectManager.class);
+		Project mockProject = mock(Project.class);
+		StageListener dummyListener = mock(StageListener.class);
+
+		try (MockedStatic<ProjectManager> pmMock = mockStatic(ProjectManager.class);
+			 MockedStatic<RequiresPermissionTask> rpMock = mockStatic(RequiresPermissionTask.class);
+			 MockedStatic<StageResourceHolder> srhMock = mockStatic(StageResourceHolder.class)) {
```
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply catroid/src/test/java/org/catrobat/catroid/stage/StageLifeCycleControllerTest.java]: I’ve addressed the changes in commit [1e53152](https://github.com/Catrobat/Catroid/pull/5164/commits/1e531528c478921b28900a6cc7cbc24f5731d1db)

File: catroid/src/test/java/org/catrobat/catroid/stage/StageLifeCycleControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+					.thenReturn(Collections.emptyList());
+			rpMock.when(() -> RequiresPermissionTask.checkPermission(any(), any()))
+					.thenReturn(true);
+
+			StageActivity.stageListener = null;
+
+			StageLifeCycleController.stageDestroy(mockStageActivity);
+
+			verify(dummyListener, never()).finish();
+			verify(mockStageActivity).manageLoadAndFinish();
```
[the remark this answers is not in the document]
Comment (wslany) [reply catroid/src/test/java/org/catrobat/catroid/stage/StageLifeCycleControllerTest.java]: Codex evaluated Copilot’s comments as valid:

dummyListener is just a local mock in the test and is never assigned to StageActivity.stageListener, which is the value stageDestroy() actually reads. Because of that, verify(dummyListener, never()).finish() will always pass regardless of the production code, so it doesn’t prove the intended behavior.

Codex evaluates them like this:

- The first comment is correct: testStageDestroySkipsListenerFinishWhenNull() is redundant or at least poorly targeted in its current form.
- The second comment is also correct: verify(dummyListener, never()).finish() is ineffective and gives false confidence.

A better version of that test would either:

- remove the third test entirely, since testStageDestroyCallsManageLoadAndFinishWhenStageListenerIsNull() already covers the important null-listener path, or
- rewrite it to assert only meaningful outcomes when StageActivity.stageListener = null, such as:
- stageDestroy() does not throw
- manageLoadAndFinish() is still called
- StageActivity.stageListener remains null

**Review by wslany — CHANGES_REQUESTED** [review]

File: catroid/src/test/java/org/catrobat/catroid/stage/StageLifeCycleControllerTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply catroid/src/test/java/org/catrobat/catroid/stage/StageLifeCycleControllerTest.java]: Okay, I’ll make these changes

File: catroid/src/test/java/org/catrobat/catroid/stage/StageLifeCycleControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+			rpMock.when(() -> RequiresPermissionTask.checkPermission(any(), any()))
+					.thenReturn(true);
+
+			StageActivity.stageListener = null;
+
+			StageLifeCycleController.stageDestroy(mockStageActivity);
+
+			verify(mockStageActivity).manageLoadAndFinish();
+			assertNull(StageActivity.stageListener);
+		}
```
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply catroid/src/test/java/org/catrobat/catroid/stage/StageLifeCycleControllerTest.java]: Done!

File: catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java [hunk 5]
```diff
@@ -259,7 +259,9 @@ static void stageDestroy(StageActivity stageActivity) {
 			if (ProjectManager.getInstance().getCurrentProject().isCastProject()) {
 				CastManager.getInstance().onStageDestroyed();
 			}
-			StageActivity.stageListener.finish();
+			if (StageActivity.stageListener != null) {
+				StageActivity.stageListener.finish();
+			}
```
[the remark this answers is not in the document]
Comment (dorianpercic) [reply catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java]: @harshsomankar123-tech please have a look.

**Review by dorianpercic — CHANGES_REQUESTED** [review]
@harshsomankar123-tech since it's a new file, could you refactor StageLifeCycleControllerTest.java into Kotlin? Since the goal is to refactor the whole project to Kotlin.

File: catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/stage/StageLifeCycleController.java]: Ohk

**Review by dorianpercic — APPROVED** [review]

**Review by wslany — APPROVED** [review]
LGTM! Thanks a lot!
